x/posts/types/models.go

```diff
@@ -0,0 +1,669 @@
+func NewEntities(hashtags []Tag, mentions []Tag, urls []Url) *Entities {
+	return &Entities{
+		Hashtags: hashtags,
+		Mentions: mentions,
+		Urls:     urls,
+	}
+}
+
+// Validate implements fmt.Validator
+func (e *Entities) Validate() error {
+	for _, tag := range e.Hashtags {
+		err := tag.Validate()
+		if err != nil {
+			return fmt.Errorf("invalid hashtag: %s", err)
+		}
+	}
+
+	for _, tag := range e.Mentions {
+		err := tag.Validate()
+		if err != nil {
+			return fmt.Errorf("invalid mention: %s", err)
+		}
+	}
+
+	for _, url := range e.Urls {
+		err := url.Validate()
+		if err != nil {
+			return fmt.Errorf("invalid url: %s", err)
+		}
+	}
+
+	// --- Make sure there are no overlapping entities based on (start, end) ---
+
+	// Map all entities into segments
+	segments := e.getSegments()
+
+	if len(segments) < 1 {
+		// We cannot have an empty entities here
+		return fmt.Errorf("entities must have at least one entity inside")
+	}
+
+	if len(segments) < 2 {
+		// With less than 2 segments there cannot be any overlap
+		return nil
+	}
+
+	// Sort the segments
+	sort.Slice(segments, func(i, j int) bool {
+		return segments[i].start < segments[j].end
+	})
+
+	// Verify there are no overlapping segments
+	for index := 0; index < len(segments)-1; index++ {
+		first, second := segments[index], segments[index+1]
+		if first.end >= second.start {
+			return fmt.Errorf("entities cannot overlap: start %d end %d", first.end, second.start)
+		}
+	}
+
+	return nil
+}
+
+type entitySegment struct {
+	start uint64
+	end   uint64
+}
+
+func (e *Entities) getSegments() []entitySegment {
+	segments := make([]entitySegment, len(e.Hashtags)+len(e.Mentions)+len(e.Urls))
+	i := 0
+	for _, hashtag := range e.Hashtags {
+		segments[i] = entitySegment{start: hashtag.Start, end: hashtag.End}
+		i++
+	}
+
+	for _, mention := range e.Mentions {
+		segments[i] = entitySegment{start: mention.Start, end: mention.End}
+		i++
+	}
+
+	for _, url := range e.Urls {
+		segments[i] = entitySegment{start: url.Start, end: url.End}
+		i++
+	}
+
+	return segments
+}
+
+// NewTag returns a new Tag instance
+func NewTag(start, end uint64, tag string) Tag {
+	return Tag{
+		Start: start,
+		End:   end,
+		Tag:   tag,
+	}
+}
+
+// Validate implements fmt.Validator
+func (t Tag) Validate() error {
+	if t.Start > t.End {
+		return fmt.Errorf("invalid start and end indexes: %d %d", t.Start, t.End)
+	}
+
+	if strings.TrimSpace(t.Tag) == "" {
+		return fmt.Errorf("invalid start and end indexes: %d %d", t.Start, t.End)
+	}
+
+	return nil
+}
+
+// NewURL returns a new Url instance
+func NewURL(start, end uint64, url, displayURL string) Url {
+	return Url{
+		Start:      start,
+		End:        end,
+		Url:        url,
+		DisplayUrl: displayURL,
+	}
+}
+
+// Validate implements fmt.Validator
+func (u Url) Validate() error {
+	if u.Start > u.End {
+		return fmt.Errorf("invalid start and end indexes: %d %d", u.Start, u.End)
+	}
+
+	if strings.TrimSpace(u.Url) == "" {
```

Comment:
shouldn't we check also that a URL starts with http/https to ensure its validity?

Reply:
I didn't add this check because there might be other different protocols that are valid and would require too much work to support. An example is the `ftp` protocol which could be used if someone hosts a file on an FTP server. Or other custom schemes that might be used by different file storages. For this reason, I think it's easier to simply leave validity checks to the DApps and allow them to store whatever URL they prefer here